trie.cpp

```diff
@@ -0,0 +1,78 @@
+#include "trie.h"
+#include <algorithm>
+
+void Trie::insertStringInternal(shared_ptr<Node> node, string s, std::size_t strIdx){
```

Comment:
I tend to not use 'using' except in very local contexts, like within a function or even a block. The reason is that it's not obvious where a name comes from otherwise. Boost is particularly bad (good?) because they have super namespace nesting and I hate it when people put a using boost::this::and::that at the top of their 2000 line source file.

Now, you might say that 'std' is an exception to that rule since it should be obvious since it's from the language itself. The problem is, I hate exceptions to rules. Just be consistent and it makes things clear and easy. Also, for things like shared_ptr that could be boost::shared_ptr or std::shared_ptr, it's just safer. Also, 'std::' isn't super hard to type nor is it very large.